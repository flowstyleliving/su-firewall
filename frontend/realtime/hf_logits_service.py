#!/usr/bin/env python3
import os
import json
from typing import List, Optional

import torch
from fastapi import FastAPI
from pydantic import BaseModel
from transformers import AutoTokenizer, AutoModelForCausalLM

app = FastAPI(title="HF Logits Service")

MODEL_CACHE = {}

def load_model(model_id: str):
	if model_id in MODEL_CACHE:
		return MODEL_CACHE[model_id]
	tokenizer = AutoTokenizer.from_pretrained(model_id)
	model = AutoModelForCausalLM.from_pretrained(model_id)
	model.eval()
	MODEL_CACHE[model_id] = (tokenizer, model)
	return tokenizer, model

class LogitsRequest(BaseModel):
	model_id: str
	prompt: str
	max_tokens: int = 32
	temperature: float = 1.0
	return_gradients: bool = False
	forced_output: Optional[str] = None
	# Sparse top-k controls
	sparse_topk: bool = False
	top_k: int = 256

@app.post("/logits")
def logits(req: LogitsRequest):
	tokenizer, model = load_model(req.model_id)
	prompt_text = req.prompt
	forced = req.forced_output
	enc_prompt = tokenizer(prompt_text, return_tensors="pt").to(model.device)

	with torch.no_grad():
		prompt_logits = model(**enc_prompt).logits  # [1, seq, vocab]
		prompt_next = prompt_logits[:, -1, :].squeeze(0)

	def softmax(x: torch.Tensor) -> torch.Tensor:
		xm = x.max().item()
		exp = torch.exp(x - xm)
		return exp / exp.sum().clamp_min(1e-30)

	def topk_probs(x: torch.Tensor, k: int):
		p = softmax(x)
		val, idx = torch.topk(p, k=min(k, p.shape[-1]))
		rest = (1.0 - val.sum().item())
		return idx.tolist(), val.tolist(), max(0.0, float(rest))

	resp = {
		"temperature": req.temperature,
	}

	# Prompt side outputs
	if req.sparse_topk:
		pi, pv, prest = topk_probs(prompt_next, req.top_k)
		resp["prompt_next_topk_indices"] = pi
		resp["prompt_next_topk_probs"] = pv
		resp["prompt_next_rest_mass"] = prest
	else:
		resp["prompt_next_logits"] = prompt_next.detach().cpu().tolist()

	token_logits: List[List[float]] = []
	topk_indices: List[List[int]] = []
	topk_probs_list: List[List[float]] = []
	rest_masses: List[float] = []

	if forced:
		enc_forced = tokenizer(forced, add_special_tokens=False, return_tensors="pt").to(model.device)
		input_ids = torch.cat([enc_prompt.input_ids, enc_forced.input_ids], dim=1)
		attention_mask = torch.cat([enc_prompt.attention_mask, enc_forced.attention_mask], dim=1)
		with torch.no_grad():
			out = model(input_ids=input_ids, attention_mask=attention_mask)
			logits_full = out.logits
			forced_len = enc_forced.input_ids.shape[1]
			start_idx = logits_full.shape[1] - forced_len
			for pos in range(start_idx, logits_full.shape[1]):
				vec = logits_full[:, pos - 1, :].squeeze(0) if pos > 0 else logits_full[:, 0, :].squeeze(0)
				if req.sparse_topk:
					idx, val, rest = topk_probs(vec, req.top_k)
					topk_indices.append(idx)
					topk_probs_list.append(val)
					rest_masses.append(rest)
				else:
					token_logits.append(vec.detach().cpu().tolist())
	else:
		vec = prompt_next
		if req.sparse_topk:
			idx, val, rest = topk_probs(vec, req.top_k)
			topk_indices.append(idx)
			topk_probs_list.append(val)
			rest_masses.append(rest)
		else:
			token_logits.append(vec.detach().cpu().tolist())

	# Build vocab if dense
	if not req.sparse_topk:
		vocab = [tokenizer.convert_ids_to_tokens(i) for i in range(model.config.vocab_size)]
		resp["vocab"] = vocab
		resp["token_logits"] = token_logits
	else:
		resp["topk_indices"] = topk_indices
		resp["topk_probs"] = topk_probs_list
		resp["rest_mass"] = rest_masses
		resp["vocab_size"] = int(model.config.vocab_size)

	return resp

if __name__ == "__main__":
	import uvicorn
	host = os.environ.get("HF_SERVICE_HOST", "127.0.0.1")
	port = int(os.environ.get("HF_SERVICE_PORT", "8899"))
	uvicorn.run("hf_logits_service:app", host=host, port=port, reload=False, workers=1) 