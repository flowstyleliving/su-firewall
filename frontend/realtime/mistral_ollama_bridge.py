#!/usr/bin/env python3
import sys
import argparse
import json
import time
import math
import requests
from typing import Dict, List, Optional

ENGINE_BASE = "http://127.0.0.1:3000/api/v1"
OLLAMA_URL = "http://127.0.0.1:11434/api/generate"


def softmax_from_logprobs(logprobs: Dict[str, float], top_k_floor: float = 1e-9) -> Dict[str, float]:
    if not logprobs:
        return {}
    # logprobs may be in natural log or log10; Ollama returns natural log base e
    # Compute normalized probabilities from provided logprobs (top-k only)
    max_logp = max(logprobs.values())
    exp_vals = {tok: math.exp(lp - max_logp) for tok, lp in logprobs.items()}
    s = sum(exp_vals.values()) or 1.0
    probs = {tok: max(top_k_floor, v / s) for tok, v in exp_vals.items()}
    # renormalize after floor
    s2 = sum(probs.values()) or 1.0
    probs = {tok: v / s2 for tok, v in probs.items()}
    return probs


def shannon_entropy_from_probs(probs: Dict[str, float]) -> float:
    if not probs:
        return 0.0
    # entropy in nats
    return -sum(p * math.log(max(p, 1e-12)) for p in probs.values())


def js_divergence(p: Dict[str, float], q: Dict[str, float]) -> float:
    # operate on union of keys
    keys = set(p.keys()) | set(q.keys())
    if not keys:
        return 0.0
    def _v(d, k):
        return d.get(k, 0.0)
    m = {k: 0.5 * (_v(p,k) + _v(q,k)) for k in keys}
    def _kl(a, b):
        s = 0.0
        for k in keys:
            ak = max(_v(a,k), 1e-12)
            bk = max(_v(b,k), 1e-12)
            s += ak * math.log(ak / bk)
        return s
    return 0.5 * (_kl(p, m) + _kl(q, m))


def post_token_metrics(session_id: str, idx: int, token: str, prob: float, entropy: float, jsd: float) -> None:
    payload = {
        "token_index": idx,
        "token_text": token,
        "probability": float(max(prob, 1e-12)),
        "entropy": float(max(entropy, 0.0)),
        "jsd": float(max(jsd, 0.0)),
    }
    url = f"{ENGINE_BASE}/session/{session_id}/token_metrics"
    try:
        requests.post(url, json=payload, timeout=5)
    except Exception as e:
        print(f"[bridge] warn: failed posting metrics: {e}", file=sys.stderr)


def stream_from_ollama(model: str, prompt: str, temperature: float, top_p: float, max_tokens: int, repetition_penalty: float, repeat_last_n: int):
    # Request streaming generation with logprobs
    req = {
        "model": model,
        "prompt": prompt,
        "stream": True,
        "options": {
            "temperature": temperature,
            "top_p": top_p,
            "num_predict": max_tokens,
            "logprobs": 20,
            "repeat_penalty": repetition_penalty,
            "repeat_last_n": repeat_last_n
        }
    }
    with requests.post(OLLAMA_URL, json=req, stream=True) as r:
        r.raise_for_status()
        for line in r.iter_lines(decode_unicode=True):
            if not line:
                continue
            try:
                yield json.loads(line)
            except Exception:
                continue


def run(session_id: str, prompt: str, model: str, temperature: float, top_p: float, max_tokens: int, repetition_penalty: float, repeat_last_n: int):
    prev_probs: Optional[Dict[str, float]] = None
    idx = 0
    # Accumulate probs for current token if provided across chunks (Ollama may send per-token)
    for chunk in stream_from_ollama(model, prompt, temperature, top_p, max_tokens, repetition_penalty, repeat_last_n):
        if chunk.get("done"):
            break
        token = chunk.get("response", "")
        # Ollama may include logprobs structure under 'logprobs' or 'top_logprobs'
        # Expect format: { "logprobs": { "<token>": logp, ... } } or a list; handle both gracefully
        lp_map: Dict[str, float] = {}
        if isinstance(chunk.get("logprobs"), dict):
            lp_map = chunk["logprobs"]
        elif isinstance(chunk.get("logprobs"), list) and chunk["logprobs"]:
            # list of {token, logprob}
            for entry in chunk["logprobs"]:
                tok = entry.get("token")
                lp = entry.get("logprob")
                if tok is not None and lp is not None:
                    lp_map[tok] = lp
        elif isinstance(chunk.get("top_logprobs"), list) and chunk["top_logprobs"]:
            for entry in chunk["top_logprobs"]:
                if isinstance(entry, dict):
                    lp_map.update({k: v for k, v in entry.items() if isinstance(v, (int, float))})
        # Build approx distribution from top-k
        probs_map = softmax_from_logprobs(lp_map) if lp_map else {}
        # Observed prob estimate
        observed_prob = probs_map.get(token, None)
        if observed_prob is None:
            # If not available, fallback to a small probability mass for the observed token
            observed_prob = 1e-3
        # Entropy and JSD
        entropy = shannon_entropy_from_probs(probs_map)
        jsd = js_divergence(prev_probs or {}, probs_map or {}) if prev_probs is not None else 0.0
        prev_probs = probs_map or prev_probs
        post_token_metrics(session_id, idx, token, observed_prob, entropy, jsd)
        idx += 1


def main():
    ap = argparse.ArgumentParser(description="Ollama â†’ Engine bridge for real inference")
    ap.add_argument("session_id", help="Existing session id from the engine")
    ap.add_argument("prompt", help="Prompt text")
    ap.add_argument("--model", default="mistral:7b", help="Ollama model name (e.g., mistral:7b)")
    ap.add_argument("--temperature", type=float, default=0.7)
    ap.add_argument("--top_p", type=float, default=0.9)
    ap.add_argument("--max_tokens", type=int, default=128)
    ap.add_argument("--repetition_penalty", type=float, default=1.15)
    ap.add_argument("--repeat_last_n", type=int, default=256)
    args = ap.parse_args()

    print(f"[bridge] starting with session={args.session_id} model={args.model}")
    print("[bridge] note: ensure ollama is running: ollama run mistral")
    try:
        run(args.session_id, args.prompt, args.model, args.temperature, args.top_p, args.max_tokens, args.repetition_penalty, args.repeat_last_n)
    except KeyboardInterrupt:
        pass


if __name__ == "__main__":
    main() 