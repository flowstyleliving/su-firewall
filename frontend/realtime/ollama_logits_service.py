#!/usr/bin/env python3
import os
import json
import hashlib
from typing import Dict, List, Tuple

import httpx
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI(title="Ollama TopK Service")

OLLAMA_URL = os.environ.get("OLLAMA_BASE_URL", "http://127.0.0.1:11434")

class OllamaTopkRequest(BaseModel):
	model_id: str
	prompt: str
	top_k: int = 128
	num_samples: int = 64
	temperature: float = 1.0
	forced_output: str | None = None

class OllamaTopkResponse(BaseModel):
	prompt_next_topk_indices: List[int]
	prompt_next_topk_probs: List[float]
	prompt_next_topk_rest_mass: float
	topk_indices: List[int]
	topk_probs: List[float]
	rest_mass: float


def _stable_token_id(token_text: str) -> int:
	# Use sha1 hash to create a stable 31-bit id
	h = hashlib.sha1(token_text.encode("utf-8")).digest()
	return int.from_bytes(h[:4], "big") & 0x7fffffff


def _sample_next_token(client: httpx.Client, model: str, prompt: str, temperature: float, seed: int) -> str:
	payload = {
		"model": model,
		"prompt": prompt,
		"stream": False,
		"raw": True,
		"options": {
			"num_predict": 1,
			"temperature": max(0.0, float(temperature)),
			"seed": int(seed),
		}
	}
	r = client.post(f"{OLLAMA_URL}/api/generate", json=payload, timeout=120)
	r.raise_for_status()
	data = r.json()
	return data.get("response", "")


def _estimate_topk(tokens: Dict[str, int], k: int) -> Tuple[List[int], List[float], float]:
	total = sum(tokens.values())
	if total <= 0:
		return [], [], 0.0
	items = sorted(tokens.items(), key=lambda x: x[1], reverse=True)
	top = items[: k]
	idx = [_stable_token_id(t) for (t, _) in top]
	probs = [cnt / total for (_, cnt) in top]
	s = float(sum(probs))
	rest = max(0.0, 1.0 - s)
	return idx, probs, rest


@app.post("/ollama_topk", response_model=OllamaTopkResponse)
def ollama_topk(req: OllamaTopkRequest):
	client = httpx.Client()
	try:
		counts_p: Dict[str, int] = {}
		counts_q: Dict[str, int] = {}
		prompt_p = req.prompt
		prompt_q = (req.prompt + " " + req.forced_output) if req.forced_output else req.prompt
		for i in range(req.num_samples):
			seed = 12345 + i
			tok_p = _sample_next_token(client, req.model_id, prompt_p, req.temperature, seed)
			counts_p[tok_p] = counts_p.get(tok_p, 0) + 1
			tok_q = _sample_next_token(client, req.model_id, prompt_q, req.temperature, seed + 9999)
			counts_q[tok_q] = counts_q.get(tok_q, 0) + 1

		pi, pv, prest = _estimate_topk(counts_p, req.top_k)
		qi, qv, qrest = _estimate_topk(counts_q, req.top_k)
		return OllamaTopkResponse(
			prompt_next_topk_indices=pi,
			prompt_next_topk_probs=pv,
			prompt_next_topk_rest_mass=prest,
			topk_indices=qi,
			topk_probs=qv,
			rest_mass=qrest,
		)
	finally:
		client.close()

if __name__ == "__main__":
	import uvicorn
	host = os.environ.get("OLLAMA_SERVICE_HOST", "127.0.0.1")
	port = int(os.environ.get("OLLAMA_SERVICE_PORT", "8898"))
	uvicorn.run("ollama_logits_service:app", host=host, port=port, reload=False, workers=1) 