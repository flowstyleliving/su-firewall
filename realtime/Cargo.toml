[package]
name = "realtime"
version = "0.1.0"
edition = "2021"

[lib]
name = "realtime"
path = "src/lib.rs"

[dependencies]
common = { path = "../common" }
preprompt = { path = "../preprompt" }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
chrono = { version = "0.4", features = ["serde"] }
anyhow = "1.0"
tracing = "0.1"
log = "0.4"
thiserror = "1.0"
ndarray = { version = "0.15", default-features = false }
axum = { version = "0.7", features = ["ws"], optional = true }
tokio = { version = "1.0", features = ["macros", "time", "sync", "rt", "rt-multi-thread"], default-features = false }
uuid = { version = "1.0", features = ["v4", "serde"] }
lru = "0.12"

[features]
default = []
api = ["axum"]
http-api = []
websocket = []
python = []
huggingface = []
llamacpp = []
pytorch = []

[dev-dependencies]
insta = { version = "1", features = ["redactions", "json"] }
tower = { version = "0.5", features = ["util"] }
http-body-util = "0.1"
hyper = "1" 