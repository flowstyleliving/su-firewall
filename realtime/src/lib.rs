pub use common::{RiskLevel, SemanticUncertaintyResult, CalibrationMode, RequestId};

pub mod live_response_auditor;
pub mod audit_interface;
pub mod live_audit_dashboard;
pub mod monitoring;
pub mod scalar_firewall;
pub mod scalar_walk_firewall;
pub mod oss_logit_adapter;
pub mod mistral_integration;
pub mod alias_ambiguity_defense;

pub mod metrics;

#[cfg(feature = "api")]
pub mod api;

#[cfg(feature = "api")]
pub use api::router; 