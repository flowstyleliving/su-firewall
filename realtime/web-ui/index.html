<!doctype html>
<html>
<head>
	<meta charset="utf-8" />
	<meta name="viewport" content="width=device-width, initial-scale=1" />
	<title>Realtime ‚Ñè_s Dashboard</title>
</head>
<body>
	<div id="root"></div>
	<script type="module">
		import init from "./realtime-web-ui.js";
		init();
	</script>
</body>
</html> 