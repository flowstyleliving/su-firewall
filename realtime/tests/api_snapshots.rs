use axum::{Router, body::Body, http::{Request, StatusCode}};
use realtime::router;
use http_body_util::BodyExt;
use tower::util::ServiceExt; // for .oneshot

fn test_router() -> Router {
	router()
}

#[tokio::test]
async fn snapshot_analyze_topk() {
	let app = test_router();
	let payload = serde_json::json!({
		"model_id": "test-model",
		"topk_indices": [[0,1,2,3,4]],
		"topk_probs": [[0.4, 0.2, 0.15, 0.15, 0.1]],
		"rest_mass": [0.0],
		"prompt_next_topk_indices": [0,1,2,3,4],
		"prompt_next_topk_probs": [0.5, 0.2, 0.1, 0.1, 0.1],
		"prompt_next_rest_mass": 0.0,
		"vocab_size": 10,
		"method": "full_fim_dir"
	});
	let req = Request::builder()
		.method("POST")
		.uri("/api/v1/analyze_topk")
		.header("content-type", "application/json")
		.body(Body::from(serde_json::to_vec(&payload).unwrap()))
		.unwrap();

	let res = app.clone().oneshot(req).await.unwrap();
	assert_eq!(res.status(), StatusCode::OK);
	let body_bytes = res.into_body().collect().await.unwrap().to_bytes();
	let mut json: serde_json::Value = serde_json::from_slice(&body_bytes).unwrap();
	// Redact non-deterministic fields
	if let Some(obj) = json.as_object_mut() {
		obj.insert("request_id".to_string(), serde_json::json!("<redacted>"));
		obj.insert("timestamp".to_string(), serde_json::json!("<redacted>"));
		obj.insert("processing_time_ms".to_string(), serde_json::json!(0.0));
	}
	insta::assert_json_snapshot!("analyze_topk_ok", json);
}

#[tokio::test]
async fn snapshot_analyze_topk_compact() {
	let app = test_router();
	let payload = serde_json::json!({
		"model_id": "test-model",
		"prompt_next_topk_indices": [0,1,2,3,4],
		"prompt_next_topk_probs": [0.5, 0.2, 0.1, 0.1, 0.1],
		"prompt_next_rest_mass": 0.0,
		"topk_indices": [0,1,2,3,4],
		"topk_probs": [0.4, 0.2, 0.15, 0.15, 0.1],
		"rest_mass": 0.0,
		"method": "full_fim_dir"
	});
	let req = Request::builder()
		.method("POST")
		.uri("/api/v1/analyze_topk_compact")
		.header("content-type", "application/json")
		.body(Body::from(serde_json::to_vec(&payload).unwrap()))
		.unwrap();

	let res = app.clone().oneshot(req).await.unwrap();
	assert_eq!(res.status(), StatusCode::OK);
	let body_bytes = res.into_body().collect().await.unwrap().to_bytes();
	let mut json: serde_json::Value = serde_json::from_slice(&body_bytes).unwrap();
	// Redact non-deterministic fields
	if let Some(obj) = json.as_object_mut() {
		obj.insert("request_id".to_string(), serde_json::json!("<redacted>"));
		obj.insert("timestamp".to_string(), serde_json::json!("<redacted>"));
		obj.insert("processing_time_ms".to_string(), serde_json::json!(0.0));
	}
	insta::assert_json_snapshot!("analyze_topk_compact_ok", json);
} 