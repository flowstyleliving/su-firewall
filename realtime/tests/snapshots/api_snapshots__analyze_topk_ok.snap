---
source: realtime/tests/api_snapshots.rs
expression: json
---
{
  "delta_mu": 0.003900000000000009,
  "delta_sigma": 0.26457513110645897,
  "free_energy": 0.9630082441355572,
  "hbar_s": 0.032122313293335404,
  "method": "full_fim_dir",
  "model_id": "test-model",
  "p_fail": 0.007849784577996727,
  "processing_time_ms": 0.0,
  "request_id": "<redacted>",
  "timestamp": "<redacted>"
}
